homeassistant/components/idasen_desk/__init__.py

```diff
@@ -68,20 +70,24 @@ async def async_disconnect(self) -> None:
         self._connection_lost = False
         await self.desk.disconnect()
 
-    @callback
-    def async_set_updated_data(self, data: int | None) -> None:
-        """Handle data update."""
+    async def async_ensure_connection_state(self) -> None:
+        """Check if the expected connection state matches the current state and correct it if needed."""
         if self._expected_connected:
             if not self.desk.is_connected:
-                _LOGGER.debug("Desk disconnected. Reconnecting")
+                _LOGGER.warning("Desk disconnected. Reconnecting")
                 self._connection_lost = True
-                self.hass.async_create_task(self.async_connect(), eager_start=False)
+                await self.async_connect()
             elif self._connection_lost:
                 _LOGGER.info("Reconnected to desk")
                 self._connection_lost = False
         elif self.desk.is_connected:
             _LOGGER.warning("Desk is connected but should not be. Disconnecting")
-            self.hass.async_create_task(self.desk.disconnect())
+            await self.desk.disconnect()
```

Comment:
It would be much cleaner if this didn't exist in the integration and the library handled this code and managed the connection state.  Right now it seems like its possible for HA and the library to disagree on which state the connection should be in

Reply:
I agree with you 100%. I was already planing on refactoring this to move that to the lib after fixing the instability. 

Regarding your second point, they can't really "disagree" since this controls whether or not it should be connected, while the lib only handles retries during the connection process, not after (it does not handle connection drops). Also, if this here tells it to disconnect, the lib loop stops, so there should be no mismatch.

Since this fixes the issues most people have been complaining about without introducing a big refactor, are you OK with getting this in first?
After that, I will work on cleaning the integration (I aim to get it to Gold🥇, eventually).

Reply:
I'm a bit worried about the potential to build up a lot of tasks from `async_set_updated_data`. Runaway task creation is a common reason we see crashes so this pattern always worries me when I see it in integrations.

If we end up with the bluetooth adv changing multiple times and `self. _expected_connected` is `False` and `self.desk.is_connected` is True, we could end up with multiple calls to `await self.desk.disconnect()` which can take a while to complete (sometimes up to 15s)

Its hard for me to tell how likely that is to happen.

If you add an `asyncio.Lock` in `async_ensure_connection_state` where it bails if the lock is already locked, I'd feel a lot more comfortable about merging it as-is

Reply:
> If you add an `asyncio.Lock` in `async_ensure_connection_state` where it bails if the lock is already locked, I'd feel a lot more comfortable about merging it as-is

If it bails out when the Lock is locked, isn't a simple boolean better?

Reply:
The Boolean flag would work as well. The lock is a bit simpler to write though since it's a context manager and you don't have to write a try/finally out to handle the case where the code in the block might fail

Reply:
Yeah, I went for the context manager since it makes it simpler.